### PR TITLE
A few more things in Cognizance

### PR DESCRIPTION
**Pull request by warp-core**
**Bugfix:**

#7026, https://steamcommunity.com/app/404410/discussions/1/3780247517054458227/

## Fix Details
- It is possible to get the first Tech Retrieval mission after having already Taely (either in Face to Maw 2 or Cognizance 11. The Tech Retrieval mission assumes you have not already met her, and Cognizance 11 assumes you have already met her. This PR adds branches for the missed cases to those two missions.
- The Expanded Horizons missions can be offered on Ssil Vida, but they seem to be written for the primary Remnant worlds. Added `remnant primary` attribute requirement to the source filter. #7026
- Cognizance 19 assumes you have already met Dawn, but it's possible to get there without meeting her. Since Cognizance 19 takes place after a bit of a climax in those missions and leads to a change of pace, it seems like a good break point so that mission now requires the player to have done Expanded Horizons Astral 2. This ensures that the player has met Dawn and that the Research Laboratory is available, which is a requirement for the next mission.
- Since Cognizance 19 can now not be done without the Research Laboratory being available, Cognizance 20 no longer needs to check for this. The blocked message was already inaccurate since the outfit is not required to start the mission, it only needs to be available, and the way the condition is used means that, if it is not available, the mission will not be instantiated and so the blocked message will not ever appear anyway.
- ~Removed the blocked message from Cognizance 32 and moved the license requirement back to the to offer condition set (from to accept). This is because, since Cognizance 19 requires some of the Astral Horizons missions, the player will have to have finished Salvage 2. The capital license is granted at the end of Salvage 3, which the player is likely to have done if they've done Salvage 2 anyway. If they haven't, it is also offered from Viminal so preventing Cognizance 32 from being instantiated will means Salvage 3 will be offered if it has not been done, failed, or aborted already. If it is done, Cognizance 32 will offfer. If it is failed or aborted, then the blocked message from Cognizance 32 isn't useful anyway.~

An alternative to requiring Expanded Horizons Astral 2 for Cognizance 19 could e to add a branch introducing Dawn and then have Cognizance 20 grant access to the Research Laboratory.

~Similarly, Cognizance 32 could grant the capital license itself, or just not require it, though that would require some adjustments to Salvage 3. This would make it possible to do all of Cognizance without doing the Salvage chain, and the addition of new salvage at the end of Cognizance might seem a bit odd if you haven't done the Salvage chain. Similarly, the Salvage chain itself will need further modification if salvage unlocked through Cognizance was available first.~

## Testing Done
0

## Save File
This save file can be used to verify the bugfix. The bug will occur when using {{insert commit hash / version}}, and will not occur when using this branch's build.
Cognizance 10 active:
[warp 0.10.1-alpha~~previous-55.txt](https://github.com/endless-sky/endless-sky/files/11647721/warp.0.10.1-alpha.previous-55.txt)

Cognizance 18 active:
[warp 0.10.1-alpha~~previous-45.txt](https://github.com/endless-sky/endless-sky/files/11647726/warp.0.10.1-alpha.previous-45.txt)

### CODE REVIEW

**Comment by Zitchas** [comment]
I haven't looked at everything yet, but just a note regarding checks for the capital license: Checks for it occur at various places intentionally as a deliberate block in the event that the player has lost the capital license, most likely due to betraying the Remnant to the Deep or any other faction that gets espionage missions against the Remnant. As such, I would prefer it if the checks for it remain, even if they are at the current time redundant.

**Comment by warp-core** [comment]
> I haven't looked at everything yet, but just a note regarding checks for the capital license: Checks for it occur at various places intentionally as a deliberate block in the event that the player has lost the capital license, most likely due to betraying the Remnant to the Deep or any other faction that gets espionage missions against the Remnant. As such, I would prefer it if the checks for it remain, even if they are at the current time redundant.

In that case, anywhere that check occurs will need to make it clear to the player that they can't actually continue if they have previously acquired the license and then lost it, without any way to recover it.

File: data/remnant/remnant 1 introduction.txt
Comment (Zitchas) [inline data/remnant/remnant 1 introduction.txt:2015]:
```suggestion
```

Comment (Zitchas) [reply data/remnant/remnant 1 introduction.txt:2015]: Instead of creating a whole new mission for this, it would be more efficient to just use actions inside the branches. The branch for already having met Taely would have just this log entry, and the branch for meeting Taely could set both, and pre-emptively set the Intro Log as being offered and done.

File: data/remnant/remnant 1 introduction.txt
Comment (Zitchas) [inline data/remnant/remnant 1 introduction.txt:1699]:
```suggestion
			`As you enter the spaceport, you notice Taely across the plaza. She begins approaching you immediately, appearing to have been waiting for you. "Greetings, <first>. I hope you are not too busy as I have an opportunity for you." When you express interest, she leads you to a balcony overlooking the primary shipyards. She seems oblivious to the chilling breeze as she leans on the railing. "As you may have noticed in your time with us, many of our ships, while they do serve us well, are old designs. While we have recently made significant progress with a new concept, we still need new ideas and inspiration." She pauses, focusing briefly on a small bay where several people appear to be pulling what looks like a supercapacitor out of a Puffin.`
```

missing quotation mark.

File: data/remnant/remnant 1 introduction.txt
Comment (Zitchas) [inline data/remnant/remnant 1 introduction.txt:3738]:
```suggestion
```

There are plenty of hallways on both Esquiline and Ssil Vida. There's no need to restrict where she might appear for this mission.

File: data/remnant/remnant 2 cognizance.txt
Comment (Zitchas) [inline data/remnant/remnant 2 cognizance.txt:1363]: This line is intentionally here so that players know that yes, there is a mission here that they can't get right now, and what's missing.

I don't see any reason for not telling them that they're missing out on something here.

File: data/remnant/remnant 2 cognizance.txt
Comment (Zitchas) [inline data/remnant/remnant 2 cognizance.txt:1367]: Does this not make it so that the mission won't even trigger at all if they don't have the capital license? 

This was intentionally setup this way to block people who have lost their capital license, and to tell them that this is what is happening. (otherwise we'd just get everyone who has lost their license posting "I'm stuck and this mission won't trigger" posts. with this `to accept` and `blocked` sections, players can put things together for themselves and realize that betraying the Remnant and losing their capital license has in fact cost them the ability to continue the campaign.

**Review by Zitchas — DISMISSED** [review]
Some good improvements. Thanks!

However also some parts that miss the mark. I have provided suggestions where I can. GitHub doesn't allow suggestions for undeleting lines, though.

File: data/remnant/remnant 1 introduction.txt
Comment (warp-core) [reply data/remnant/remnant 1 introduction.txt:2015]: I don't think it's more efficient to duplicate the same log text that is made in the intro log mission elsewhere.
If that is more efficient though, should I also be deleting the other log missions you have here and applying their logs at the appropriate locations in mission conversations as well?

Comment (Zitchas) [reply data/remnant/remnant 1 introduction.txt:2015]: Sorry, I was referring specifically to the Penguin log entry. The other entry can be triggered by various places, so having a single mission for it is easier. Although that being said, ensuring the penguin entry comes after the first one could be problematic then, so maybe the way you have it is better, unless there's some way to specify which order they are in.

File: data/remnant/remnant 1 introduction.txt
Comment (warp-core) [reply data/remnant/remnant 1 introduction.txt:3738]: It might be useful for you to comment on #7026 if you don't think that's actually a problem.

File: data/remnant/remnant 2 cognizance.txt
Comment (warp-core) [reply data/remnant/remnant 2 cognizance.txt:1363]: As it is, the only way for this mission to be blocked is if the player dosn't have the remnant capital license, which will only be the case if they haven't gotten far enough in the Salvage chain, in which case this mission should just not attempt to offer and instead those missions can be offered.

This situation may warrant a more complex way of guiding the player, though.

File: data/remnant/remnant 2 cognizance.txt
Comment (warp-core) [reply data/remnant/remnant 2 cognizance.txt:1367]: I'm having difficulty finding anything in the game that will actually take away the player's remnant capital license.

Regardless, using a blocked message to indicate that the player is permanently locked out of continuing these missions because of something they did before doesn't seem very good. That message will appear every time the player comes here and seems to imply that they need to do something to meet the mission requirements (which is what every other blocked message in the game does) when, in fact, there's nothing they can do short of editing the files.

**Comment by warp-core** [comment]
I've reverted the changes to Cognizance 32 because, looking at it again, I think it needs more changes to work better.

**Review by bene-dictator — APPROVED** [review]
I think this looks fine, waiting on Zitchas' review.

File: data/remnant/remnant 2 cognizance.txt
Comment (warp-core) [reply data/remnant/remnant 2 cognizance.txt:704]: This message is shown when the player meets the "to offer" conditions, but cannot accept the mission.
Insufficient cargo space is a way in which that situation could arise, so it makes sense to mention that here.
Not having a research laboratory installed, on the other hand, will not actually lead to this message appearing. You don't need to have a research laboratory to be offered this mission. So, it doesn't make sense to say you need one in the blocked message.

**Comment by tibetiroka** [comment]
@Zitchas Are you going to review this any time soon? It seems to be waiting on your approval.

File: data/remnant/remnant 1 introduction.txt
Comment (tibetiroka) [inline data/remnant/remnant 1 introduction.txt:1640]: Should this not be called `remnant: met taely`?

Comment (warp-core) [reply data/remnant/remnant 1 introduction.txt:1640]: This is the name of the condition that is currently used. Perhaps it should be renamed to that, but I don't think that's a question this PR needs to answer.

File: data/remnant/remnant 1 introduction.txt
Comment (roadrunner56) [inline data/remnant/remnant 1 introduction.txt:1964]:
```suggestion
		log "People" "Taely" `One of Taely's main priorities is improving the maneuverability characteristics of Remnant ships. To this end she has been working on building a new prototype ship based on the void sprites.`
```

File: data/remnant/remnant 2 cognizance.txt
Comment (roadrunner56) [inline data/remnant/remnant 2 cognizance.txt:446]: Feels like this reads better.
```suggestion
			`On Viminal, you easily locate the maintenance yards. You ask around for Prefect Taely and are eventually directed towards a solidly built woman working on a Pelican. She spots you approaching and frees herself from the machinery. "Greetings, Captain <last>, I am Prefect Taely. I understand you have something for me." You hand her the data crystal. "This must contain the detailed chemical analysis and profiles," she says. She slots it into her console and taps a few commands. "Okay, I just need to incorporate these profiles into the system and make a few more adjustments. Meet me in the spaceport cafeteria in a couple hours."`
```

File: data/remnant/remnant 2 cognizance.txt
Comment (roadrunner56) [inline data/remnant/remnant 2 cognizance.txt:449]:
```suggestion
			`On Viminal, you easily locate the maintenance yards and eventually find Taely buried in a Pelican. She spots you approaching and frees herself from the machinery. "Greetings, <first>. How is our station faring?" You reply that it is improving rapidly and hand her the data crystal. "Ah, good. This must contain the detailed chemical analysis and profiles," she says. She slots it into her console and taps a few commands. "Okay, I just need to incorporate these profiles into the system and make a few more adjustments. Meet me in the spaceport cafeteria in a couple hours."`
```

File: data/remnant/remnant 1 introduction.txt
Comment (tibetiroka) [reply data/remnant/remnant 1 introduction.txt:1640]: Since the condition is incremented directly before the conversation, is it not safe to assume that the player has it? If not, the incrementing should be done after the branch check.

File: data/remnant/remnant 2 cognizance.txt
Comment (tibetiroka) [inline data/remnant/remnant 2 cognizance.txt:447]: The same thing about the incrementing and branching. Doesn't seem to make sense.

**Review by tibetiroka — COMMENTED** [review]
Cognizance 18 could use an `on visit` to tell you that your escorts haven't arrived in the system yet. There is also a typo in that mission; `onscreen` instead of `on screen`.

File: data/remnant/remnant 1 introduction.txt
Comment (warp-core) [reply data/remnant/remnant 1 introduction.txt:1640]: The content of a mission action is not necessarily executed in the order it is listed.
The relevant part of this here is that a conversation panel is constructed before any condition changes in the mission action are applied. During construction, the conversation panel finds the first text it should display, which involves evaluating and following any branches it finds prior to visible text.
Here, that means this branch will check the condition before the increment operation is applied.

**Comment by warp-core** [comment]
> Cognizance 18 could use an `on visit` to tell you that your escorts haven't arrived in the system yet. There is also a typo in that mission; `onscreen` instead of `on screen`.

Doesn't appear to be related to the changes in this PR?

**Comment by tibetiroka** [comment]
Yeah, it's fine to do that in another PR. I just noticed it because that's the mission your save started with.

**Review by tibetiroka — APPROVED** [review]

**Comment by roadrunner56** [comment]
@warp-core Can you fix the merge conflicts on this PR?

**Pull request merged**
